### PR TITLE
Autolog from Keras to MLflow

### PR DESCRIPTION
**Pull request by apurva-koti**
## What changes are proposed in this pull request?
 
As in #1520, MLflow users can now use `mlflow.keras.autolog()` to automatically capture metrics, parameters and tags from a keras training session without the need for manual log statements.
 
## How is this patch tested?
 
Unit testing
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
Use `mlflow.keras.autolog()` to automatically log accuracy, loss, optimizer data and post-training model checkpointing from your Keras (note: not `tensorflow.keras`) training session.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [X] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/keras.py [hunk 1]
```diff
@@ -311,3 +316,64 @@ def load_model(model_uri, **kwargs):
...
+
+
+class __MLflowKerasCallback(Callback):
+    """
+    Callback for auto-logging metrics and parameters.
+    Records available logs after each epoch.
+    Records model structural information as params after training finishes.
+    """
+
+    def __enter__(self):
```
Comment (smurching) [inline mlflow/keras.py]: Do we actually need the context manager methods (`__enter__` and `__exit__`)? Doesn't look like they do anything

File: mlflow/keras.py [hunk 2]
```diff
@@ -311,3 +316,64 @@ def load_model(model_uri, **kwargs):
...
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    def on_epoch_end(self, epoch, logs=None):
+        if not logs:
+            return
+        try:
+            mlflow.log_metrics(logs, step=epoch)
```
Comment (smurching) [inline mlflow/keras.py]: Should we log metrics etc asynchronously with Keras too?

File: mlflow/keras.py (hunk 1 above)
Comment (apurva-koti) [reply mlflow/keras.py]: True, removed

File: mlflow/keras.py [hunk 3]
```diff
@@ -13,6 +13,10 @@
 import importlib
 import os
 import yaml
+import gorilla
+import warnings
+from keras.callbacks import Callback
```
Comment (smurching) [inline mlflow/keras.py]: Ah I remember why we don't want to import keras at the top-level here - it's so that `mlflow.keras.log_model` can be called on a `tf.keras` model with just TensorFlow installed (& not Keras). So we should probably put our import under `__MLflowKerasCallback`

**Review by smurching — COMMENTED** [review]
Mostly looks great! Just a couple of comments

File: mlflow/keras.py (hunk 2 above)
Comment (smurching) [reply mlflow/keras.py]: Discussed offline - we'll perf test this & add the async stuff if we need to.

File: mlflow/keras.py [hunk 4]
```diff
@@ -311,3 +316,64 @@ def load_model(model_uri, **kwargs):
...
+            sum_list = []
+            self.model.summary(print_fn=sum_list.append)
+            summary = '\n'.join(sum_list)
+            mlflow.set_tag('summary', summary)
+            log_model(self.model, artifact_path='model')
+        except mlflow.exceptions.MlflowException as e:
+            warnings.warn("Logging to Mlflow failed: " + str(e))
+
+
+def autolog():
```
Comment (smurching) [inline mlflow/keras.py]: Add experimental annotation?

Comment (smurching) [reply mlflow/keras.py]: Would also be nice to flesh out this docstring a bit more (e.g. the TF docstring has):
```
    """
    Enable automatic logging from TensorFlow to MLflow. If applicable,
    model checkpoints are logged as artifacts to a 'models' directory, along
    with any TensorBoard log data.

    Refer to the tracking documentation for
    information on what is logged with different TensorFlow workflows.

    :param metrics_every_n_steps: The frequency with which metrics should be logged.
                                  Defaults to 100. Ex: a value of 100 will log metrics
                                  at step 0, 100, 200, etc.

    """
```

Would be good to replicate that, maybe we can actually include the URL of the tracking docs in^ (e.g. https://mlflow.org/docs/latest/tracking.html)

**Review by smurching — APPROVED** [review]
LGTM thanks for the updates!

**Comment by mobiuscreek** [comment]
Thanks for adding this functionality! Currently the keras autologging function only applies to the _fit_ function of keras. Is there a plan to extend it to _fit_generator_ ?  It still needs a custom callback to make it work.
